alonzo/impl/src/Cardano/Ledger/Alonzo/Rules/Utxow.hs

```diff
@@ -0,0 +1,277 @@
+import Cardano.Ledger.Alonzo.Scripts (Script)
+import Cardano.Ledger.Alonzo.Tx
+  ( ScriptPurpose,
+    Tx,
+    checkScriptData,
+    hashWitnessPPData,
+    isNonNativeScriptAddress,
+    language,
+    scriptsNeeded,
+    wits',
+  )
+import Cardano.Ledger.Alonzo.TxBody (WitnessPPDataHash)
+import Cardano.Ledger.Alonzo.TxWitness (TxWitness (..))
+import qualified Cardano.Ledger.Core as Core
+import Cardano.Ledger.Era (Crypto, Era, ValidateScript (..))
+import Cardano.Ledger.SafeHash (EraIndependentData, SafeHash)
+import Control.Iterate.SetAlgebra (domain, eval, (◁))
+import Control.State.Transition.Extended
+import Data.Coders
+import qualified Data.Map.Strict as Map
+import Data.Set (Set)
+import qualified Data.Set as Set
+import Data.Typeable (Typeable)
+import GHC.Records
+import Shelley.Spec.Ledger.BaseTypes
+  ( ShelleyBase,
+    StrictMaybe (..),
+  )
+import Shelley.Spec.Ledger.LedgerState (UTxOState (..))
+import Shelley.Spec.Ledger.STS.Utxo (UtxoEnv (..))
+import Shelley.Spec.Ledger.STS.Utxow
+  ( ShelleyStyleWitnessNeeds,
+    UtxowPredicateFailure (..),
+    shelleyStyleWitness,
+  )
+import Shelley.Spec.Ledger.Scripts (ScriptHash (..))
+import Shelley.Spec.Ledger.Tx (TxIn (..))
+
+-- =====================================================
+
+-- | The Predicate failure type in the Alonzo Era. It embeds the Predicate
+--   failure type of the Shelley Era, as they share some failure modes.
+data AlonzoPredFail era
+  = Embed (UtxowPredicateFailure era)
+  | UnRedeemableScripts [(ScriptPurpose (Crypto era), ScriptHash (Crypto era))]
+  | MissingNeededScriptHash (Set (ScriptHash (Crypto era)))
+  | DataHashSetsDontAgree
+      (Set (DataHash (Crypto era)))
+      -- ^ from the Tx
+      (Set (DataHash (Crypto era)))
+      -- ^ from the UTxO restricted to the Tx inputs
+  | PPViewHashesDontMatch
+      (StrictMaybe (WitnessPPDataHash (Crypto era)))
+      -- ^ The PPHash in the TxBody
+      (StrictMaybe (WitnessPPDataHash (Crypto era)))
+      -- ^ Computed from the current Protocol Parameters
+
+deriving instance
+  ( Era era,
+    Show (PredicateFailure (Core.EraRule "UTXO" era)) -- The Shelley UtxowPredicateFailure needs this to Show
+  ) =>
+  Show (AlonzoPredFail era)
+
+deriving instance
+  ( Era era,
+    Eq (PredicateFailure (Core.EraRule "UTXO" era)) -- The Shelley UtxowPredicateFailure needs this to Eq
+  ) =>
+  Eq (AlonzoPredFail era)
+
+instance
+  ( Era era,
+    ToCBOR (PredicateFailure (Core.EraRule "UTXO" era)),
+    Typeable (Core.AuxiliaryData era),
+    Typeable (Core.Script era)
+  ) =>
+  ToCBOR (AlonzoPredFail era)
+  where
+  toCBOR x = encode (encodePredFail x)
+
+encodePredFail ::
+  ( Era era,
+    ToCBOR (PredicateFailure (Core.EraRule "UTXO" era)),
+    Typeable (Core.Script era),
+    Typeable (Core.AuxiliaryData era)
+  ) =>
+  AlonzoPredFail era ->
+  Encode 'Open (AlonzoPredFail era)
+encodePredFail (Embed x) = Sum Embed 0 !> E toCBOR x
```

Comment:
This is a bit involved, since the CBOR instances for (UtxowPredicateFailure era) in the Shelley era (ie. the type of x in (Embed x)) is written by hand, so making it inline would require going back and changing that. For a future time?  Even if we flattened in CBOR, it remains unflattened (and visible) in the type AlonzoPredFail.